Analysis/Utkscan/processors/source/PidProcessor.cpp

```diff
@@ -27,325 +27,896 @@ using namespace std;
+   DeclareHistogram2D(DD_TOF1_PIN0_GATED_FIT_FLIP, SB, SD, "FIT: Flipped Tof1 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF1_PIN0_GATED_YSO_FLIP, SB, SD, "YSO: Flipped Tof1 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF1_PIN0_GATED_RIT_FLIP, SB, SD, "RIT: Flipped Tof1 vs Pin0 dE ");
+
+   DeclareHistogram2D(DD_TOF2_PIN0, SB, SD, "Tof2 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF2_PIN0_GATED_FIT, SB, SD, "FIT: Tof2 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF2_PIN0_GATED_YSO, SB, SD, "YSO: Tof2 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF2_PIN0_GATED_RIT, SB, SD, "RIT: Tof2 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF2_PIN0_FLIP, SB, SD, "Flipped Tof2 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF2_PIN0_GATED_FIT_FLIP, SB, SD, "FIT: Flipped Tof2 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF2_PIN0_GATED_YSO_FLIP, SB, SD, "YSO: Flipped Tof2 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF2_PIN0_GATED_RIT_FLIP, SB, SD, "RIT: Flipped Tof2 vs Pin0 dE ");
+
+   DeclareHistogram2D(DD_TOF3_PIN0, SB, SD, "Tof3 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF3_PIN0_GATED_FIT, SB, SD, "FIT: Tof3 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF3_PIN0_GATED_YSO, SB, SD, "YSO: Tof3 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF3_PIN0_GATED_RIT, SB, SD, "RIT: Tof3 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF3_PIN0_FLIP, SB, SD, "Flipped Tof3 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF3_PIN0_GATED_FIT_FLIP, SB, SD, "FIT: Flipped Tof3 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF3_PIN0_GATED_YSO_FLIP, SB, SD, "YSO: Flipped Tof3 vs Pin0 dE ");
+   DeclareHistogram2D(DD_TOF3_PIN0_GATED_RIT_FLIP, SB, SD, "RIT: Flipped Tof3 vs Pin0 dE ");
+
+   DeclareHistogram1D(D_RANGE, S3, "Range distribution with PID gate");
+
+   DeclareHistogram2D(DD_PIN2_3, SC, SC, "Pin2 vs Pin3 dE ");
+   DeclareHistogram2D(DD_PIN0_2, SC, SC, "Pin0 vs Pin2 dE ");
+   DeclareHistogram2D(DD_PIN1_3, SC, SC, "Pin1 vs Pin3 dE ");
+   DeclareHistogram2D(DD_TOF4_PIN2, SB, SD, "Tof4 vs Pin2 dE ");
+   DeclareHistogram2D(DD_TOF5_PIN2, SB, SD, "Tof5 vs Pin2 dE ");
+   DeclareHistogram2D(DD_TOF4_PIN3, SB, SD, "Tof4 vs Pin3 dE ");
+   DeclareHistogram2D(DD_TOF5_PIN3, SB, SD, "Tof5 vs Pin3 dE ");
 }  // Declare plots
 
-PidProcessor::PidProcessor(const double &YSO_Implant_thresh, const double &FIT_thresh) : EventProcessor(OFFSET, RANGE, "PidProcessor") {
-    associatedTypes.insert("pid");
-    associatedTypes.insert("pin");
+PidProcessor::PidProcessor(const double &YSO_Implant_thresh, const double &FIT_thresh, const double &RIT_thresh) : EventProcessor(OFFSET, RANGE, "PidProcessor") {
+   associatedTypes.insert("pid");
+   associatedTypes.insert("pin");
+
+   yso_threshold_ = YSO_Implant_thresh;
+   fit_threshold_ = FIT_thresh;
+   rit_threshold_ = RIT_thresh;
 
-    yso_threshold_ = YSO_Implant_thresh;
-    fit_threshold_ = FIT_thresh;
+   yso_energy_prev = 0;
+   fit_energy_prev = 0;
+   fit_energy_prev = 0;
 }
 
 bool PidProcessor::PreProcess(RawEvent &event) {
```

Comment:
Are we really worried about imageL/R, dispL/R and such being nullptr in this function? If this is a concern, then each time we grab them from the vectors, for instance in this block
```cpp
if(!imageL_vec.empty() ){
  auto imageL = std::max_element(imageL_vec.begin(), imageL_vec.end(), compare_energy);
  double internalTAC_Convert_Tick_adc = Globals::get()->GetAdcClockInSeconds((*imageL)->GetChanID().GetModFreq()) * 1e9;
  if((*imageL)){
    pid_struct.image_scint_L_logic_time = (*imageL)->GetTime() * internalTAC_Convert_Tick_adc;
  }
}
```
we need to put anything using (*image) behind a guard like I've done. It seems like the checks for nullptr might not be neccessary, based on the code getting this far.
```cpp
if(!imageL_vec.empty() ){
  auto imageL = std::max_element(imageL_vec.begin(), imageL_vec.end(), compare_energy);
  if((*imageL)){
    double internalTAC_Convert_Tick_adc = Globals::get()->GetAdcClockInSeconds((*imageL)->GetChanID().GetModFreq()) * 1e9;
    pid_struct.image_scint_L_logic_time = (*imageL)->GetTime() * internalTAC_Convert_Tick_adc;
  }
}
```

Reply:
GetSummary can return a null DetectorSummary if the object doesn't exist already. So the calls significantly earlier when we do `event.GetSummary("pid:cross_scint_b1", true)->GetList();`, but the true second argument causes the summary to be constructed assuming the xmldriver hasn't already made the summary. But yeah according to DetectorDriver.cpp, if the list isn't empty then the objects are never null. This will take a while to fix appropriately.

Reply:
Should just merge it as it is now as the check is basically free and this should be addressed in a different PR/issue